dm/master/workerrpc/rawgrpc.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package workerrpc
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/errors"
+	"google.golang.org/grpc"
+
+	"github.com/pingcap/dm/dm/pb"
+)
+
+// GRPCClient stores raw grpc connection and worker client
+type GRPCClient struct {
+	conn   *grpc.ClientConn
+	client pb.WorkerClient
+	closed int32
+}
+
+// NewGRPCClientWrap initializes a new grpc client from given grpc connection and worker client
+func NewGRPCClientWrap(conn *grpc.ClientConn, client pb.WorkerClient) (*GRPCClient, error) {
+	return &GRPCClient{
+		conn:   conn,
+		client: client,
+		closed: 0,
+	}, nil
+}
+
+// NewGRPCClient initializes a new grpc client from worker address
+func NewGRPCClient(addr string) (*GRPCClient, error) {
+	conn, err := grpc.Dial(addr, grpc.WithInsecure(), grpc.WithBackoffMaxDelay(3*time.Second))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return NewGRPCClientWrap(conn, pb.NewWorkerClient(conn))
+}
+
+// SendRequest implements Client.SendRequest
+func (c *GRPCClient) SendRequest(ctx context.Context, req *Request, timeout time.Duration) (*Response, error) {
+	if atomic.LoadInt32(&c.closed) != 0 {
+		return nil, errors.New("send request on a closed client")
+	}
+	if req.Type != CmdFetchDDLInfo {
+		ctx1, cancel := context.WithTimeout(ctx, timeout)
+		defer cancel()
+		return callRPC(ctx1, c.client, req)
+	}
+	return callRPC(ctx, c.client, req)
+}
+
+// Close implements Client.Close
+func (c *GRPCClient) Close() error {
+	defer atomic.CompareAndSwapInt32(&c.closed, 0, 1)
+	if c.conn == nil {
+		return nil
+	}
+	err := c.conn.Close()
```

Comment:
Is `c.conn.Close()` idempotent?

Reply:
not, add `c.conn = nil` in defer function to ensure `c.conn` is nil after Close